### ISSUE TITLE
Changing document ID in fscrawler

### ISSUE THREAD

**Issue by Jasmeet2011** [dadoonet/fscrawler#1432]
We are indexing several folders many of which have duplicate files. We would like to index only one copy of the duplicate files in the index. 
One way which ES provides is to write a separate query in logstash for deduplication.
 Is it possible to change the document id value to checksum value at the time of writing to ES index? This will help in preventing writing of duplicate files to the index.

**Comment by dadoonet** [comment]
That's already the case but the folder is taken into account as well.

So we could add an option to ignore the folder name may be.

**Comment by Jasmeet2011** [comment]
Could you please tell whether there are any settings that can be done in the yaml file to change the document id to a user defined field (checksum in this case).
I could not locate it in the documentation. 
Thanks

**Comment by dadoonet** [comment]
Actually, I think you could use https://fscrawler.readthedocs.io/en/fscrawler-2.9/admin/fs/local-fs.html#using-filename-as-elasticsearch-id

**Comment by dadoonet** [comment]
The same filename will generate only one document.

**Comment by dadoonet** [comment]
Do you mean checksum on the binary content itself?

**Comment by Jasmeet2011** [comment]
yes, that way i can prevent duplicate files getting indexed from the folders which contain duplicates.

**Comment by dadoonet** [comment]
So we do have the [feature `fs.checksum`](https://fscrawler.readthedocs.io/en/latest/admin/fs/local-fs.html#file-checksum) which generates a checksum of the binary file within the field `file.checksum`.

I think that an ingest pipeline could cover this:

```
PUT _ingest/pipeline/set-id-from-checksum
{
  "description": "Set the _id from file.checksum",
  "processors": [
    {
      "set": {
        "field": "_id",
        "value": "{{{file.checksum}}}"
      }
    }
  ]
}
```

Then you can define this `set-id-from-checksum` in [`elasticsearch.pipeline`](https://fscrawler.readthedocs.io/en/latest/admin/fs/elasticsearch.html#using-ingest-node-pipeline).

Note that the fingerprint is computed from the binary file, not from the extracted text. If you want to do this on the extracted text, you can compute a fingerprint using this time the [`fingerprint` ingest processor](https://www.elastic.co/docs/reference/ingest-processor/fingerprint-processor):

```
PUT _ingest/pipeline/content-fingerprint-id
{
  "description": "Compute a tmp fingerprint from the content field and set it as the document _id",
  "processors": [
    {
      "fingerprint": {
        "fields": ["content"],
        "target_field": "_tmp_fingerprint",
        "method": "SHA-256"
      }
    },
    {
      "set": {
        "field": "_id",
        "value": "{{{_tmp_fingerprint}}}"
      }
    },
    {
      "remove": {
        "field": "_tmp_fingerprint",
        "ignore_missing": true
      }
    }
  ]
}
```

Note that this last option will erase documents with no extracted text or the exact same text content.

Knowing the workaround, I don't think we should add this into FSCrawler but document it.

**Comment by dadoonet** [comment]
Note that this has been documented at https://fscrawler.readthedocs.io/en/latest/user/tips.html#deduplicate-documents-with-a-content-based-id
